### PR TITLE
[WIP] feat: proxyevm module

### PR DESCRIPTION
**Pull request by JoowonYun**
## Effects

- [ ] Soft update
- [x] Breaking change
- [ ] Chain fork related

## Major Reviewer
@yoosah 

## Checklist

- [ ] Backward compatible?
- [ ] Test enough in your local environment?
- [ ] Add related test cases?

### CODE REVIEW

File: proto/xpla/proxyevm/v1beta1/tx.proto [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import "ethermint/evm/v1/tx.proto";
+
+service Msg {
+  // FundFeeCollector defines a method to allow an account to directly
+  // fund the fee collector.
+  rpc CallEVM(MsgCallEVM) returns (ethermint.evm.v1.MsgEthereumTxResponse);
+}
+
+// MsgCallEVM submits the given message data to a evm smart contract
+message MsgCallEVM {
```
Comment (yoosah) [inline proto/xpla/proxyevm/v1beta1/tx.proto:17]: XXXRequest is better I though.

File: x/proxyevm/keeper/evm.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	gasCap := config.DefaultGasCap
+
+	nonce, err := k.accountKeeper.GetSequence(ctx, sender)
+	if err != nil {
+		return nil, gasCap, err
+	}
+
+	from := common.BytesToAddress(sender.Bytes())
+
+	if commit {
```
Comment (yoosah) [inline x/proxyevm/keeper/evm.go:33]: remember this is always true currently

File: x/proxyevm/keeper/msg_server.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+	for _, coin := range msg.Funds {
+		if coin.Denom == evmParams.EvmDenom {
+			fundAmount = coin.Amount
+		}
+	}
+
+	res, gasLimit, err := k.callEVM(
+		ctx,
+		sender,
+		contract,
```
Comment (yoosah) [inline x/proxyevm/keeper/msg_server.go:67]: null check?

File: x/proxyevm/keeper/msg_server.go [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+	var labels []metrics.Label
+	sender, err := sdk.AccAddressFromBech32(msg.Sender)
+	if err != nil {
+		return nil, err
+	}
+
+	var contract *common.Address
+	if msg.Contract != "" {
+		addr := common.HexToAddress(msg.Contract)
+		contract = &addr
```
Comment (yoosah) [inline x/proxyevm/keeper/msg_server.go:45]: check local reference

File: x/proxyevm/keeper/msg_server.go [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+var _ types.MsgServer = msgServer{}
+
+func (k msgServer) CallEVM(goCtx context.Context, msg *types.MsgCallEVM) (*evmtypes.MsgEthereumTxResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	txIndex := k.evmKeeper.GetTxIndexTransient(ctx)
+
+	var labels []metrics.Label
+	sender, err := sdk.AccAddressFromBech32(msg.Sender)
```
Comment (yoosah) [inline x/proxyevm/keeper/msg_server.go:37]: think about address check position

File: x/reward/handler.go [hunk 6]
```diff
@@ -20,7 +20,7 @@ func NewHandler(k keeper.Keeper) sdk.Handler {
 			return sdk.WrapServiceResult(ctx, res, err)
 
 		default:
-			return nil, sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "unrecognized distribution message type: %T", msg)
+			return nil, sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "unrecognized reward message type: %T", msg)
```
Comment (yoosah) [inline x/reward/handler.go:23]: divide commit pls

**Comment by psy2848048** [comment]
Plain ERC20 token transfer consumes 1.2M gas units, while direct EVM call consumes 300K

Test code: https://github.com/xpladev/xpla/blob/test/proxyevm/integration_test/integration_test.go#L795-L892
Solidity bytecode: https://github.com/xpladev/xpla/blob/test/proxyevm/integration_test/misc/token.sol.bin

**Comment by psy2848048** [comment]
Update

- Gas usage captured in Antehandler as: https://github.com/xpladev/xpla/commit/2c9bb1cfdd0c22d1c5cc79432794ffb5f33e2f7d
- Almost no difference in Antehandler
  - Tx from EVM directly (thru 8545) vs  Tx from Proxy EVM directly (thru 9090)
```plain
EVM ante handler
SetUpContext consumed 0
MempoolFee consumed 0
MinGasPrice consumed 7702
ValidateBasic consumed 23093
SigVerification consumed 30782
AccVerification consumed 30782
GasConsume consumed 61075
CanTransfer consumed 84168
GasConsume 2 consumed 114536
Increment sender seq consumed 123278
GasWanted consumed 138669
Event emission consumed 138769
```

```plain
Cosm ante handler
Limit simulator consumed 0
Count tx consumed 3429
RejectExtentionOption consumed 3429
MempoolFee consumed 3429
Min gas fee consumed 11131
ValidateBasic consumed 11131
Tx timeout height consumed 11131
ValidateMemo consumed 16533
Consume Gas for tx size consumed 25895
Deduct Fee consumed 40797
Set pub key consumed 52029
Validator sig count consumed 57431
Sig gas consume consumed 85604
Gas fee check 2 consumed 87375
Inc fee check consumed 98856
Ibc ante fee check consumed 98856
Gas fee check 3 consumed 114247
```

- With the low gas limit (around 300000), it's enough in Antehandler, so the tx submited
- But the gas deplete when EVM call complete and processing with the state db
- Debug trace:
```plain
github.com/cosmos/cosmos-sdk/baseapp.newOutOfGasRecoveryMiddleware.func1
    github.com/cosmos/cosmos-sdk@v0.45.9/baseapp/recovery.go:55
github.com/cosmos/cosmos-sdk/baseapp.newRecoveryMiddleware.func1
    github.com/cosmos/cosmos-sdk@v0.45.9/baseapp/recovery.go:39
github.com/cosmos/cosmos-sdk/baseapp.processRecovery
    github.com/cosmos/cosmos-sdk@v0.45.9/baseapp/recovery.go:28
github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).runTx.func1
    github.com/cosmos/cosmos-sdk@v0.45.9/baseapp/baseapp.go:632
    runtime.gopanic\n\truntime/panic.go:838
github.com/cosmos/cosmos-sdk/store/types.(*basicGasMeter).ConsumeGas
    github.com/cosmos/cosmos-sdk@v0.45.9/store/types/gas.go:99
github.com/cosmos/cosmos-sdk/store/gaskv.(*Store).Get
    github.com/cosmos/cosmos-sdk@v0.45.9/store/gaskv/store.go:43
github.com/cosmos/cosmos-sdk/store/prefix.Store.Get
    github.com/cosmos/cosmos-sdk@v0.45.9/store/prefix/store.go:68.                    <- from here
github.com/cosmos/cosmos-sdk/x/params/types.Subspace.GetIfExists
    github.com/cosmos/cosmos-sdk@v0.45.9/x/params/types/subspace.go:118
github.com/evmos/ethermint/x/evm/keeper.Keeper.GetParams
    github.com/evmos/ethermint@v0.19.3/x/evm/keeper/params.go:14
github.com/xpladev/xpla/x/proxyevm/keeper.msgServer.CallEVM                             <- also from here
    github.com/xpladev/xpla/x/proxyevm/keeper/msg_server.go:55
github.com/xpladev/xpla/x/proxyevm/types._Msg_CallEVM_Handler.func1
    github.com/xpladev/xpla/x/proxyevm/types/tx.pb.go:205
github.com/cosmos/cosmos-sdk/baseapp.(*MsgServiceRouter).RegisterService.func2.1
    github.com/cosmos/cosmos-sdk@v0.45.9/baseapp/msg_service_router.go:113
github.com/xpladev/xpla/x/proxyevm/types._Msg_CallEVM_Handler
    github.com/xpladev/xpla/x/proxyevm/types/tx.pb.go:207
github.com/cosmos/cosmos-sdk/baseapp.(*MsgServiceRouter).RegisterService.func2
    github.com/cosmos/cosmos-sdk@v0.45.9/baseapp/msg_service_router.go:126
github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).runMsgs
    github.com/cosmos/cosmos-sdk@v0.45.9/baseapp/baseapp.go:760
github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).runTx
    github.com/cosmos/cosmos-sdk@v0.45.9/baseapp/baseapp.go:717
github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).DeliverTx
    github.com/cosmos/cosmos-sdk@v0.45.9/baseapp/abci.go:282
github.com/tendermint/tendermint/abci/client.(*localClient).DeliverTxAsync
    github.com/tendermint/tendermint@v0.34.21/abci/client/local_client.go:93
github.com/tendermint/tendermint/proxy.(*appConnConsensus).DeliverTxAsync
    github.com/tendermint/tendermint@v0.34.21/proxy/app_conn.go:85
github.com/tendermint/tendermint/state.execBlockOnProxyApp
    github.com/tendermint/tendermint@v0.34.21/state/execution.go:320
github.com/tendermint/tendermint/state.(*BlockExecutor).ApplyBlock
    github.com/tendermint/tendermint@v0.34.21/state/execution.go:140
github.com/tendermint/tendermint/consensus.(*State).finalizeCommit
    github.com/tendermint/tendermint@v0.34.21/consensus/state.go:1659
github.com/tendermint/tendermint/consensus.(*State).tryFinalizeCommit
    github.com/tendermint/tendermint@v0.34.21/consensus/state.go:1568
github.com/tendermint/tendermint/consensus.(*State).enterCommit.func1
    github.com/tendermint/tendermint@v0.34.21/consensus/state.go:1503
github.com/tendermint/tendermint/consensus.(*State).enterCommit
    github.com/tendermint/tendermint@v0.34.21/consensus/state.go:1541
github.com/tendermint/tendermint/consensus.(*State).addVote
    github.com/tendermint/tendermint@v0.34.21/consensus/state.go:2155
github.com/tendermint/tendermint/consensus.(*State).tryAddVote
    github.com/tendermint/tendermint@v0.34.21/consensus/state.go:1953
github.com/tendermint/tendermint/consensus.(*State).handleMsg
    github.com/tendermint/tendermint@v0.34.21/consensus/state.go:856
github.com/tendermint/tendermint/consensus.(*State).receiveRoutine
    github.com/tendermint/tendermint@v0.34.21/consensus/state.go:763

out of gas in location: ReadPerByte; gasWanted: 120000, gasUsed: 120952: out of gas
```

**Comment by psy2848048** [comment]
High gas usage occurs in `evmKeeper.EstimateGas` which only proxyEVM tx passes

- Gas usage captured in Antehandler & kpper as: https://github.com/xpladev/xpla/commit/351ed231d2486c78ed2e9b7d5c5e3d2aa092f0be
```plain
Cosm ante handler
Limit simulator consumed 0
Count tx consumed 3429
RejectExtentionOption consumed 3429
MempoolFee consumed 3429
Min gas fee consumed 11131
ValidateBasic consumed 11131
Tx timeout height consumed 11131
ValidateMemo consumed 16533
Consume Gas for tx size consumed 25905
Deduct Fee consumed 40837
Set pub key consumed 52069
Validator sig count consumed 57471
Sig gas consume consumed 85644
Gas fee check 2 consumed 87415
Inc fee check consumed 98896
Ibc ante fee check consumed 98896
Gas fee check 3 consumed 114287
keeper, GetTxIndexTransient 114387
keeper, Address processing 114387
keeper, GetParams 122076
keeper, msg.Fund processing 122076
callEVM, head of call evm 123847
callEVM, in commit 123847
callEVM, after gas estimation 1108773
callEVM, after call evm 1195750
keeper, after evm execution 1195750
keeper, after add attr1 1195750
keeper, after add attr2 1195750
keeper, after add contract res 1195750
keeper, after after event emission 1195750
```

File: proto/xpla/proxyevm/v1beta1/tx.proto (hunk 1 above)
Comment (JoowonYun) [reply proto/xpla/proxyevm/v1beta1/tx.proto:17]: followed cosmos message convention

File: x/proxyevm/keeper/evm.go (hunk 2 above)
Comment (JoowonYun) [reply x/proxyevm/keeper/evm.go:33]: Ok, I'll add Query calls later

File: x/reward/handler.go (hunk 6 above)
Comment (JoowonYun) [reply x/reward/handler.go]: https://github.com/xpladev/xpla/pull/51/commits/dc9cedb08c682d720a144670f478ce9709faa266

File: x/proxyevm/keeper/msg_server.go (hunk 5 above)
Comment (JoowonYun) [reply x/proxyevm/keeper/msg_server.go]: https://github.com/xpladev/xpla/pull/51/files#diff-eb144f638aa2d9b38e9b922e996317d447e2bd853623b03c973cce548848fbc7R37

Already checked in `ValidateBasic`. So change to must

File: x/proxyevm/keeper/msg_server.go (hunk 4 above)
Comment (JoowonYun) [reply x/proxyevm/keeper/msg_server.go]: https://github.com/xpladev/xpla/pull/51/files#diff-eb144f638aa2d9b38e9b922e996317d447e2bd853623b03c973cce548848fbc7R39

File: x/proxyevm/keeper/msg_server.go (hunk 3 above)
Comment (JoowonYun) [reply x/proxyevm/keeper/msg_server.go]: Changed to local variable.
https://github.com/xpladev/xpla/pull/51/files#diff-eb144f638aa2d9b38e9b922e996317d447e2bd853623b03c973cce548848fbc7R39

**Review by psy2848048 — APPROVED** [review]
```
======== Starting TestEVMContractAndTx/Test01_CheckBalance ... ========
======== Starting TestEVMContractAndTx/Test02_DeployTokenContract ... ========
Tx hash:  0xae3aed4e153447f9575614fac669531d92334aced0c9f9dd48e24b0c95aafc6c
Token address:  0xFdBcb2c3506586211AB194820E333191Bc242eAc
======== Starting TestEVMContractAndTx/Test03_ExecuteTokenContractAndQueryOnEvmJsonRpc ... ========
Tx hash:  0x9a87a4a2163db0a4865bcffcb8b0f3d6356928e89ad07078b4c761617e5c13ce
Gas used: 300000
Balance validated!
======== Starting TestEVMContractAndTx/Test04_ExecuteTokenContractByProxy ... ========
Tx sent 57B363E4EC840CB1BCE6B23F837F05BA416E254FFAED02817DF22FFEED8DECB9
gas used: 210781
Tx applied 57B363E4EC840CB1BCE6B23F837F05BA416E254FFAED02817DF22FFEED8DECB9
20000000000000000000
Balance validated!
PASS
ok      github.com/xpladev/xpla/integration_test        16.159s
```

👍

**Pull request closed without merging**